genie-agent/src/main/java/com/netflix/genie/agent/execution/services/NaiveFetchingCacheService.java

```diff
@@ -0,0 +1,150 @@
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.tuple.Pair;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.core.io.Resource;
+import org.springframework.core.io.ResourceLoader;
+import org.springframework.stereotype.Service;
+import org.springframework.util.DigestUtils;
+import org.springframework.util.FileCopyUtils;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.util.Set;
+
+/**
+ * A naive implementation of a file cache on local disk that uses URIs as keys and transparently downloads
+ * missing entries.
+ *
+ * Does NOT handle other agents reading/writing/deleting in the same cache folder
+ * Does NOT handle visioning or staleness of locally cached resources
+ * Does NOT handle concurrency withing the same agent (not an issue at the moment)
+ * Does NOT garbage-collect old items
+ * TODO: implement a cache that handles the above
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Service
+@Lazy
+@Slf4j
+class NaiveFetchingCacheService implements FetchingCacheService {
+
+    private final ResourceLoader resourceLoader;
+    private final File cacheDirectory;
+
+    NaiveFetchingCacheService(
+        final ResourceLoader resourceLoader,
+        final ArgumentDelegates.CacheArguments cacheArguments
+    ) {
+        this.resourceLoader = resourceLoader;
+
+        this.cacheDirectory = cacheArguments.getCacheDirectory();
+        if (!cacheDirectory.exists()) {
+            try {
+                Files.createParentDirs(new File(cacheDirectory, "_"));
+            } catch (final IOException e) {
+                throw new RuntimeException("Failed to create cache directory: " + cacheDirectory.getAbsolutePath(), e);
+            }
+        } else if (!cacheDirectory.isDirectory()) {
+            throw new RuntimeException("Cache location is not directory: " + cacheDirectory.getAbsolutePath());
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void get(final URI sourceFileUri, final File destinationFile) throws DownloadException, IOException {
+        final File cachedFile = lookupOrDownload(sourceFileUri);
+        Files.copy(cachedFile, destinationFile);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void get(final Set<Pair<URI, File>> sourceDestinationPairs) throws DownloadException, IOException {
+        for (final Pair<URI, File> sourceDestinationPair : sourceDestinationPairs) {
+            get(sourceDestinationPair.getKey(), sourceDestinationPair.getValue());
+        }
+    }
+
+    private File lookupOrDownload(final URI sourceFileUri) throws DownloadException, IOException {
+
+        final String uriString = sourceFileUri.toASCIIString();
+
+        log.info("Lookup: {}", uriString);
```

Comment:
Debug?

Reply:
I guess. Need another pass at this to add metrics anyway, going to merge as-is.